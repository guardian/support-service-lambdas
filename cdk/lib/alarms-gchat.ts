import type {
	GuLambdaErrorPercentageMonitoringProps,
	NoMonitoring,
} from '@guardian/cdk/lib/constructs/cloudwatch';
import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack, GuStringParameter } from '@guardian/cdk/lib/constructs/core';
import { GuSnsLambdaExperimental } from '@guardian/cdk/lib/experimental/patterns';
import type { App } from 'aws-cdk-lib';
import { CfnMapping, Duration } from 'aws-cdk-lib';
import { SnsTopic } from 'aws-cdk-lib/aws-events-targets';
import { Effect, Policy, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { Topic } from 'aws-cdk-lib/aws-sns';

export interface AlarmsGchatProps extends GuStackProps {
	stack: string;
	stage: string;
}

export class AlarmsGchat extends GuStack {
	constructor(scope: App, id: string, props: AlarmsGchatProps) {
		super(scope, id, props);

		const app = 'alarms-gchat';
		const nameWithStage = `${app}-${this.stage}`;

		const commonEnvironmentVariables = {
			App: app,
			Stack: this.stack,
			Stage: this.stage,
		};

		const parameters = {
			webhook: new GuStringParameter(this, 'webhook', {
				description: 'Google Chat webhook for lambda to fire to',
			}),
		};
		// const stageMapping = new CfnMapping(this, 'stageMapping', {
		// 	mapping: {
		// 		CODE: { alarmActionsEnabled: 'FALSE' },
		// 		PROD: { alarmActionsEnabled: 'TRUE' },
		// 	},
		// });

		const snsTopic = new Topic(this, `${app}-topic`, {
			topicName: `${app}-topic-${props.stage}`,
		});

		// const getMonitoringConfiguration = ():
		// 	| NoMonitoring
		// 	| GuLambdaErrorPercentageMonitoringProps => {
		// 	if (this.stage === 'PROD' || this.stage === 'CODE') {
		// 		return {
		// 			alarmName: `Failed to raise alarm via Google Chat in ${this.stage}`,
		// 			alarmDescription:
		// 				'Please check the CloudWatch Alarms dashboard to see which alarms are firing',
		// 			toleratedErrorPercentage: 0,
		// 			snsTopicName: 'mobile-server-side-email-alert',
		// 			// We can't use ternaries or other code-based ways of doing this as we synthesise this before applying, meaning this will always return false and therefore CODE
		// 			actionsEnabled: stageMapping.findInMap(
		// 				this.stage,
		// 				'alarmActionsEnabled',
		// 			) as unknown as boolean,
		// 		};
		// 	}
		// 	return {
		// 		noMonitoring: true,
		// 	};
		// };

		const lambda = new GuSnsLambdaExperimental(
			this,
			'google-chat-bot-sns-to-gchat',
			{
				description:
					'An API Gateway triggered lambda generated in the support-service-lambdas repo',
				functionName: nameWithStage,
				fileName: `${app}.zip`,
				handler: 'index.handler',
				runtime: Runtime.NODEJS_18_X,
				memorySize: 1024,
				environment: {
					WEBHOOK: parameters.webhook.valueAsString,
					...commonEnvironmentVariables,
				},
				timeout: Duration.seconds(15),
				existingSnsTopic: { externalTopicName: snsTopic.topicName },
				monitoringConfiguration: { noMonitoring: true },
				app: app,
			},
		);

		const s3InlinePolicy: Policy = new Policy(this, 'S3 inline policy', {
			statements: [
				new PolicyStatement({
					effect: Effect.ALLOW,
					actions: ['s3:GetObject'],
					resources: [
						`arn:aws:s3::*:membership-dist/${this.stack}/${this.stage}/${app}/`,
					],
				}),
			],
		});

		lambda.role?.attachInlinePolicy(s3InlinePolicy);
	}
}
