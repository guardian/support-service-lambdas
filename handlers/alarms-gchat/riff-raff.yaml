stacks:
- support
regions:
- eu-west-1
allowedStages:
  - CODE
  - PROD
deployments:
  alarms-gchat-cloudformation:
    type: cloud-formation
    app: alarms-gchat
    parameters:
      templateStagePaths:
        CODE: alarms-gchat-CODE.template.json
        PROD: alarms-gchat-PROD.template.json

  alarms-gchat:
    type: aws-lambda
    parameters:
      fileName: alarms-gchat.zip
      bucketSsmLookup: true
      prefixStack: false
      functionNames:
      - alarms-gchat-
    dependencies: [alarms-gchat-cloudformation]
