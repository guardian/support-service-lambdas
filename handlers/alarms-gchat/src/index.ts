import type {
	Handler,
	SNSEvent
} from 'aws-lambda';

interface AlarmMessage {
	AlarmName: string;
	AlarmDescription?: string;
	NewStateReason: string;
}

export const handler: Handler = async (
	event: SNSEvent,
): Promise<null> => {
	console.log(`Input is ${JSON.stringify(event)}`);

	const webhookUrl = process.env.WEBHOOK;

	if (!webhookUrl) {
		console.error('No webhook provided');
		return null;
	}

	const messages: string[] = event.Records.map(record => {
		const message = JSON.parse(record.Sns.Message) as AlarmMessage;
		return `ALARM: ${message.AlarmName} has triggered!\n\nDescription: ${message.AlarmDescription ?? ''}\n\nReason: ${message.NewStateReason}}`
	});

	const responses = messages.map(message => {
		console.log(`Sending: ${message}`);
		return fetch(webhookUrl, {
			method: 'POST',
			headers: {'Content-Type': 'application/json'},
			body: JSON.stringify({
				text: message,
			}),
		});
	});

	return Promise.all(responses).then((responses) => {
		responses.forEach(response => {
			console.log(`Response: ${response.status}`);
		});
		return null
	})
};
