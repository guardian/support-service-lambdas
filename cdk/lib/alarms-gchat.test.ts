
import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { AlarmsGchat } from './alarms-gchat';

describe('The Alarms gchat stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const codeStack = new AlarmsGchat(app, 'alarms-gchat-CODE', {
			stack: 'membership',
			stage: 'CODE',
		});
		const prodStack = new AlarmsGchat(app, 'alarms-gchat-PROD', {
			stack: 'membership',
			stage: 'PROD',
		});

		expect(Template.fromStack(codeStack).toJSON()).toMatchSnapshot();
		expect(Template.fromStack(prodStack).toJSON()).toMatchSnapshot();
	});
});
