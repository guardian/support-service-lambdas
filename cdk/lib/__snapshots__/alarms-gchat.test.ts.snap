// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Alarms gchat stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuSnsLambdaExperimental",
    ],
    "gu:cdk:version": "TEST",
  },
  "Outputs": {
    "googlechatbotsnstogchatTopicName31621335": {
      "Value": {
        "Fn::GetAtt": [
          "alarmsgchattopic56F6C617",
          "TopicName",
        ],
      },
    },
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "S3inlinepolicy3B07399A": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": "arn:aws:s3::*:membership-dist/membership/CODE/alarms-gchat/",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "S3inlinepolicy3B07399A",
        "Roles": [
          {
            "Ref": "googlechatbotsnstogchatServiceRole6E143D03",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "alarmsgchattopic56F6C617": {
      "Properties": {
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
        "TopicName": "alarms-gchat-topic-CODE",
      },
      "Type": "AWS::SNS::Topic",
    },
    "googlechatbotsnstogchat496F36E9": {
      "DependsOn": [
        "googlechatbotsnstogchatServiceRoleDefaultPolicyF93B28A9",
        "googlechatbotsnstogchatServiceRole6E143D03",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "membership/CODE/alarms-gchat/alarms-gchat.zip",
        },
        "Description": "An API Gateway triggered lambda generated in the support-service-lambdas repo",
        "Environment": {
          "Variables": {
            "APP": "alarms-gchat",
            "App": "alarms-gchat",
            "STACK": "membership",
            "STAGE": "CODE",
            "Stack": "membership",
            "Stage": "CODE",
          },
        },
        "FunctionName": "alarms-gchat-CODE",
        "Handler": "index.handler",
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "googlechatbotsnstogchatServiceRole6E143D03",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "alarms-gchat",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
        "Timeout": 15,
      },
      "Type": "AWS::Lambda::Function",
    },
    "googlechatbotsnstogchatAllowInvokealarmsgchatCODEgooglechatbotsnstogchatSnsExistingIncomingEventsTopic4CE261FCAA86CD34": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "googlechatbotsnstogchat496F36E9",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:aws:sns:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Fn::GetAtt": [
                  "alarmsgchattopic56F6C617",
                  "TopicName",
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "googlechatbotsnstogchatServiceRole6E143D03": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "alarms-gchat",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "googlechatbotsnstogchatServiceRoleDefaultPolicyF93B28A9": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/membership/CODE/alarms-gchat/alarms-gchat.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/CODE/membership/alarms-gchat",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/CODE/membership/alarms-gchat/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "googlechatbotsnstogchatServiceRoleDefaultPolicyF93B28A9",
        "Roles": [
          {
            "Ref": "googlechatbotsnstogchatServiceRole6E143D03",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "googlechatbotsnstogchatgooglechatbotsnstogchatSnsExistingIncomingEventsTopic90B77594": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "googlechatbotsnstogchat496F36E9",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "Region": {
          "Ref": "AWS::Region",
        },
        "TopicArn": {
          "Fn::Join": [
            "",
            [
              "arn:aws:sns:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Fn::GetAtt": [
                  "alarmsgchattopic56F6C617",
                  "TopicName",
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
  },
}
`;

exports[`The Alarms gchat stack matches the snapshot 2`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuSnsLambdaExperimental",
    ],
    "gu:cdk:version": "TEST",
  },
  "Outputs": {
    "googlechatbotsnstogchatTopicName31621335": {
      "Value": {
        "Fn::GetAtt": [
          "alarmsgchattopic56F6C617",
          "TopicName",
        ],
      },
    },
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "S3inlinepolicy3B07399A": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": "arn:aws:s3::*:membership-dist/membership/PROD/alarms-gchat/",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "S3inlinepolicy3B07399A",
        "Roles": [
          {
            "Ref": "googlechatbotsnstogchatServiceRole6E143D03",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "alarmsgchattopic56F6C617": {
      "Properties": {
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "TopicName": "alarms-gchat-topic-PROD",
      },
      "Type": "AWS::SNS::Topic",
    },
    "googlechatbotsnstogchat496F36E9": {
      "DependsOn": [
        "googlechatbotsnstogchatServiceRoleDefaultPolicyF93B28A9",
        "googlechatbotsnstogchatServiceRole6E143D03",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "membership/PROD/alarms-gchat/alarms-gchat.zip",
        },
        "Description": "An API Gateway triggered lambda generated in the support-service-lambdas repo",
        "Environment": {
          "Variables": {
            "APP": "alarms-gchat",
            "App": "alarms-gchat",
            "STACK": "membership",
            "STAGE": "PROD",
            "Stack": "membership",
            "Stage": "PROD",
          },
        },
        "FunctionName": "alarms-gchat-PROD",
        "Handler": "index.handler",
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "googlechatbotsnstogchatServiceRole6E143D03",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "alarms-gchat",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "Timeout": 15,
      },
      "Type": "AWS::Lambda::Function",
    },
    "googlechatbotsnstogchatAllowInvokealarmsgchatPRODgooglechatbotsnstogchatSnsExistingIncomingEventsTopic3416B87B209BB21B": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "googlechatbotsnstogchat496F36E9",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:aws:sns:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Fn::GetAtt": [
                  "alarmsgchattopic56F6C617",
                  "TopicName",
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "googlechatbotsnstogchatServiceRole6E143D03": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "alarms-gchat",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "googlechatbotsnstogchatServiceRoleDefaultPolicyF93B28A9": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/membership/PROD/alarms-gchat/alarms-gchat.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/membership/alarms-gchat",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/membership/alarms-gchat/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "googlechatbotsnstogchatServiceRoleDefaultPolicyF93B28A9",
        "Roles": [
          {
            "Ref": "googlechatbotsnstogchatServiceRole6E143D03",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "googlechatbotsnstogchatgooglechatbotsnstogchatSnsExistingIncomingEventsTopic90B77594": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "googlechatbotsnstogchat496F36E9",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "Region": {
          "Ref": "AWS::Region",
        },
        "TopicArn": {
          "Fn::Join": [
            "",
            [
              "arn:aws:sns:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Fn::GetAtt": [
                  "alarmsgchattopic56F6C617",
                  "TopicName",
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
  },
}
`;
